### PR TITLE
11780 rebased allow reauth w qb whenever

### PR DESCRIPTION
**Pull request by DevneyHamilton**
Summary of changes in this PR:
- remove intuit's cdn js oauth flow
- add more oauth token information to connection model
- on accounting settings page, give admins oppty to reconnect to qb if refreshing the oauth2 token fails
- add feature specs for the settings show page, covering the qb controller; and model  specs on connection object
- log qb auth info when tokens are updated or refresh fails
- remove annotation gem & its annotations

### CODE REVIEW

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 1]
```diff
@@ -2,14 +2,19 @@ class Admin::Accounting::QuickbooksController < Admin::AdminController
...
-    redirect_uri = oauth_callback_admin_accounting_quickbooks_url
-    grant_url = qb_consumer.auth_code.authorize_url(
-      redirect_uri: redirect_uri,
-      response_type: "code",
-      state: SecureRandom.hex(12),
-      scope: "com.intuit.quickbooks.accounting"
-    )
-    redirect_to grant_url
+    if Rails.env.test?
+      redirect_to oauth_callback_admin_accounting_quickbooks_path(state: "test_state", realmId: "test_realm_id")
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb:6]: Very nice

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 2]
```diff
@@ -18,16 +23,22 @@ def oauth_callback
 
     # Fetch and store the access token and other necessary authentication information.
     if params[:state]
-      redirect_uri = oauth_callback_admin_accounting_quickbooks_url
-      response = qb_consumer.auth_code.get_token(params[:code], redirect_uri: redirect_uri)
+      response = oath_token_response
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: s/oath/oauth/

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 3]
```diff
@@ -71,4 +77,17 @@ def update_changed
   def qb_consumer
     @qb_consumer ||= Accounting::QB::Consumer.new
   end
+
+  def oath_token_response
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: Lovely!

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 4]
```diff
@@ -71,4 +77,17 @@ def update_changed
...
+    if Rails.env.test?
+      OpenStruct.new({
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: Don't need curlies b/c hash is last argument

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 5]
```diff
@@ -18,16 +23,22 @@ def oauth_callback
 
     # Fetch and store the access token and other necessary authentication information.
     if params[:state]
-      redirect_uri = oauth_callback_admin_accounting_quickbooks_url
-      response = qb_consumer.auth_code.get_token(params[:code], redirect_uri: redirect_uri)
+      response = oath_token_response
       if response
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: Can we please change this to raise an exception if response is nil? It really shouldn't be, right? We should be notified if it is?

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 6]
```diff
@@ -18,16 +23,22 @@ def oauth_callback
...
-        Accounting::QB::Connection.create(
+        connection_attrs = {
           access_token: response.token,
+          invalid_grant: false,
+          last_updated_at: Time.current,
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: I'm curious about this addition...

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 7]
```diff
@@ -18,16 +23,22 @@ def oauth_callback
...
           refresh_token: response.refresh_token,
           realm_id: params[:realmId],
           division: Division.root,
           token_expires_at: Time.zone.at(response.expires_at)
-        )
+        }
+        connection = Accounting::QB::Connection.first
+        connection ||= Accounting::QB::Connection.new
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: I think you can replace the above with `connection = Accounting::QB::Connection.first_or_initialize`

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 8]
```diff
@@ -18,16 +23,22 @@ def oauth_callback
...
+        connection.update(connection_attrs)
+        connection.save!
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: `update` saves. So `update!` should be sufficient on its own to run validations and save.

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 9]
```diff
@@ -18,16 +23,22 @@ def oauth_callback
...
+        connection.log_token_info("OAuth connection updated in OAuth callback")
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb]: Cool!

File: app/controllers/admin/accounting/quickbooks_controller.rb [hunk 10]
```diff
@@ -39,6 +50,7 @@ def oauth_callback
 
     flash[:notice] = t('quickbooks.connection.link_message')
     flash[:alert] = t('quickbooks.connection.import_in_progress_message')
+    redirect_to admin_accounting_settings_path
```
Comment (smoyte) [inline app/controllers/admin/accounting/quickbooks_controller.rb:53]: Yay!

File: app/models/accounting/qb/connection.rb [hunk 11]
```diff
@@ -1,61 +1,13 @@
...
-  def refresh_token!
-    qb_consumer = Accounting::QB::Consumer.new.oauth_consumer
-    oauth2_token = OAuth2::AccessToken.new(qb_consumer, self.access_token, {refresh_token: self.refresh_token})
-    refreshed = oauth2_token.refresh!.to_hash
-    self.access_token = refreshed[:access_token]
-    self.refresh_token = refreshed[:refresh_token]
-    self.token_expires_at = Time.zone.at(refreshed[:expires_at])
-    self.save!
+  def connected?
+    !expired? && !invalid_grant
```
Comment (smoyte) [inline app/models/accounting/qb/connection.rb:10]: Hooray for adding nil constraints and simplifying code!

File: app/models/accounting/qb/connection.rb [hunk 12]
```diff
@@ -76,4 +28,48 @@ def company_name
...
+      self.access_token,
+      {refresh_token: self.refresh_token}
+    )
+    begin
+      refreshed = oauth2_token.refresh!.to_hash
+    rescue OAuth2::Error
+      log_token_info("OAuth2 error when refreshing token. Old token info")
+      self.invalid_grant = true
+      return false
+    else
```
Comment (smoyte) [inline app/models/accounting/qb/connection.rb:64]: Nice use of `else` here. I had to go and look up when to use this vs. just the keeping it in the begin block :)

File: app/models/accounting/qb/connection.rb [hunk 13]
```diff
@@ -76,4 +28,48 @@ def company_name
...
+      self.access_token = refreshed[:access_token]
+      self.refresh_token = refreshed[:refresh_token]
+      self.token_expires_at = Time.zone.at(refreshed[:expires_at])
+      self.last_updated_at = Time.current
```
Comment (smoyte) [inline app/models/accounting/qb/connection.rb:68]: Also curious about this. This implies the fetch is complete but it isn't. I wonder if we should still allow NULL here?

File: app/models/accounting/qb/connection.rb [hunk 14]
```diff
@@ -76,4 +28,48 @@ def company_name
...
+      self.invalid_grant = false
+      self.save!
+      log_token_info("Successfully refreshed token")
+      return true
```
Comment (smoyte) [inline app/models/accounting/qb/connection.rb:72]: `return` is unnecessary here

File: app/models/accounting/qb/connection.rb [hunk 15]
```diff
@@ -76,4 +28,48 @@ def company_name
...
+      qb_consumer,
+      self.access_token,
+      {refresh_token: self.refresh_token}
+    )
+    begin
+      refreshed = oauth2_token.refresh!.to_hash
+    rescue OAuth2::Error
+      log_token_info("OAuth2 error when refreshing token. Old token info")
+      self.invalid_grant = true
+      return false
```
Comment (smoyte) [inline app/models/accounting/qb/connection.rb:63]: `return` is unnecessary here

File: app/models/accounting/qb/full_fetcher.rb [hunk 16]
```diff
@@ -30,7 +30,7 @@ def fetch_all
       private
 
       def fetch_qb_data
-        started_fetch_at = Time.zone.now
+        started_fetch_at = Time.current
```
Comment (smoyte) [inline app/models/accounting/qb/full_fetcher.rb:33]: Thanks for cleanup!

File: app/models/division.rb [hunk 17]
```diff
@@ -161,11 +108,11 @@ def qb_accounts_selected?
     accounts.size == 3
   end
 
-  # If no QB connection on this division, fall back to nearest ancestor with QB connection.
+  # If no valid QB connection on this division, fall back to nearest ancestor with QB connection.
   # May return nil.
   def qb_division
     # Division.root
-    qb_connection ? self : parent&.qb_division
+    qb_connection&.connected? ? self : parent&.qb_division
```
Comment (smoyte) [inline app/models/division.rb:115]: Thanks for keeping the multi-QB connections dream alive!

File: app/policies/accounting/transaction_policy.rb [hunk 18]
```diff
@@ -44,7 +44,7 @@ def read_only_reasons
   private(*delegate(:division, to: :loan))
 
   def machine_user_or_appropriate_division_admin?
-    # If qb_division is nil, it means no qb_connection has been set up on any of division's ancestors.
+    # If qb_division is nil, it means no valid qb_connection exists on any of division's ancestors.
```
Comment (smoyte) [inline app/policies/accounting/transaction_policy.rb:47]: Thanks for updating docs!

File: app/views/admin/accounting/settings/show.html.slim [hunk 19]
```diff
@@ -1,10 +1,4 @@
 - content_for(:title, t("settings.accounting"))
-
-script type="text/javascript" src="https://appcenter.intuit.com/Content/IA/intuit.ipp.anywhere-1.3.3.js"
-
-javascript:
-  intuit.ipp.anywhere.setup({menuProxy: '/path/to/blue-dot', grantUrl: '#{authenticate_admin_accounting_quickbooks_url}'});
-
```
Comment (smoyte) [inline app/views/admin/accounting/settings/show.html.slim:7]: Hooray!

File: db/migrate/20210520193148_add_invalid_grant_flag_to_connection_and_add_constraints.rb [hunk 20]
```diff
@@ -0,0 +1,13 @@
...
+  def change
+    add_column :accounting_qb_connections, :invalid_grant, :boolean, null: false, default: false
+    remove_index :accounting_qb_connections, :division_id
+    add_index :accounting_qb_connections, :division_id, unique: true
+    change_column_null :accounting_qb_connections, :access_token, false
+    change_column_null :accounting_qb_connections, :refresh_token, false
+    change_column_null :accounting_qb_connections, :token_expires_at, false
+    change_column_null :accounting_qb_connections, :division_id, false
+    change_column_null :accounting_qb_connections, :realm_id, false
+    change_column_null :accounting_qb_connections, :last_updated_at, false
```
Comment (smoyte) [inline db/migrate/20210520193148_add_invalid_grant_flag_to_connection_and_add_constraints.rb:11]: Perhaps revert this one?

File: spec/factories/accounting_qb_connections_factory.rb [hunk 21]
```diff
@@ -26,6 +26,7 @@
     division { root_division }
     realm_id { 'xxx' }
     access_token { 'xxx' }
+    last_updated_at { Time.current }
```
Comment (smoyte) [inline spec/factories/accounting_qb_connections_factory.rb:29]: This probably goes too?

File: spec/factories/divisions_factory.rb [hunk 22]
```diff
@@ -77,7 +77,11 @@ def root_division
 
     trait :with_qb_connection do
       after(:create) do |division|
-        division.qb_connection = create(:accounting_qb_connection, division: division)
+        division.qb_connection = create(
+          :accounting_qb_connection,
+          division: division,
+          last_updated_at: Time.current
+        )
```
Comment (smoyte) [inline spec/factories/divisions_factory.rb:84]: And this?

File: spec/features/admin/accounting/settings_flow_spec.rb [hunk 23]
```diff
@@ -0,0 +1,104 @@
+require "rails_helper"
+
+feature "settings flow", :accounting do
```
[the remark this answers is not in the document]
Comment (smoyte) [reply spec/features/admin/accounting/settings_flow_spec.rb:3]: Bleh we need to disable this for specs but we can do that later.

File: spec/features/admin/accounting/settings_flow_spec.rb [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+      end
+    end
+  end
+
+  describe "initial page load and authentication" do
+    context "qb connection exists but qb grant invalid" do
+      before do
+        division.qb_connection.update!(invalid_grant: true)
+      end
+      scenario do
```
Comment (smoyte) [inline spec/features/admin/accounting/settings_flow_spec.rb:34]: Space between before and scenario please.

File: spec/features/admin/accounting/settings_flow_spec.rb [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+        visit "/admin/accounting/settings"
+        prin_acct_name = accounts.first.name
+        int_rcv_acct_name = accounts[1].name
+        int_inc_acct_name = accounts[2].name
+        select prin_acct_name, from: "division[principal_account_id]"
+        select int_rcv_acct_name, from: "division[interest_receivable_account_id]"
+        select int_inc_acct_name, from: "division[interest_income_account_id]"
+        cbd = Time.zone.today.to_s
+        fill_in "Closed Books Date", with: Time.zone.today.to_s
```
Comment (smoyte) [inline spec/features/admin/accounting/settings_flow_spec.rb:74]: Use `cbd` variable here?

File: spec/features/admin/accounting/settings_flow_spec.rb [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+          job_class: "FullFetcherJob",
+          job_succeeded_at: Time.current
+        )
+      end
+
+      scenario do
+        visit "/admin/accounting/settings"
+        prin_acct_name = accounts.first.name
+        int_rcv_acct_name = accounts[1].name
+        int_inc_acct_name = accounts[2].name
```
Comment (smoyte) [inline spec/features/admin/accounting/settings_flow_spec.rb]: Can you please move these into `let`s? It makes the scenario body less heavy.

File: spec/features/admin/accounting/settings_flow_spec.rb [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+        cbd = Time.zone.today.to_s
+        fill_in "Closed Books Date", with: Time.zone.today.to_s
+        check "division[qb_read_only]"
+        click_on "Save"
+        expect(page).to have_content "Connected to "
+        expect(page).to have_select("division[principal_account_id]", selected: prin_acct_name)
+        expect(page).to have_select("division[interest_receivable_account_id]", selected: int_rcv_acct_name)
+        expect(page).to have_select("division[interest_income_account_id]", selected: int_inc_acct_name)
+        expect(page).to have_field("Closed Books Date", with: cbd)
+        expect(page).to have_checked_field("division[qb_read_only]")
```
Comment (smoyte) [inline spec/features/admin/accounting/settings_flow_spec.rb:83]: Very cool spec!

File: spec/features/admin/accounting/settings_flow_spec.rb [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+        Task.create(
+          job_type_value: :full_fetcher,
+          activity_message_value: "x",
+          job_class: "FullFetcherJob",
+          job_succeeded_at: Time.current
+        )
+      end
+
+      scenario do
+        visit "/admin/accounting/settings"
```
Comment (smoyte) [inline spec/features/admin/accounting/settings_flow_spec.rb:69]: Should we check that there is a "QuickBooks data import succeeded." message in here or something? We test for the 'pending' one elsewhere...

File: spec/features/admin/accounting/settings_flow_spec.rb [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+        expect(page).to have_content "Connected to "
+        expect(page).to have_select("division[principal_account_id]", selected: prin_acct_name)
+        expect(page).to have_select("division[interest_receivable_account_id]", selected: int_rcv_acct_name)
+        expect(page).to have_select("division[interest_income_account_id]", selected: int_inc_acct_name)
+        expect(page).to have_field("Closed Books Date", with: cbd)
+        expect(page).to have_checked_field("division[qb_read_only]")
+      end
+    end
+
+    context "connected but qb full fetch is pending" do
```
Comment (smoyte) [inline spec/features/admin/accounting/settings_flow_spec.rb:87]: This is covered above, is it not? 

Separately, should we test the case where the job fails? I think the message is 'QuickBooks data import failed'

File: spec/models/accounting/qb/connection_spec.rb [hunk 30]
```diff
@@ -1,124 +1,31 @@
...
     end
 
-    it_behaves_like 'no token is present'
-  end
-
-  context 'with invalid realm id' do
-    subject(:connection) do
-      Accounting::QB::Connection.new(token: valid_token, secret: valid_secret, realm_id: invalid_realm_id, token_expires_at: 180.days.from_now.utc)
+    it "is not connected" do
+      expect(connection.connected?).to be false
```
Comment (smoyte) [inline spec/models/accounting/qb/connection_spec.rb:28]: Hooray for simplifications!

Comment (smoyte) [reply spec/models/accounting/qb/connection_spec.rb:28]: And for removing x'd specs!

**Review by smoyte — COMMENTED** [review]
Terrific work! Big steps forward, lots of cleanups and improvements. Kudos!

Perhaps the biggest change, not big, that I'm suggesting is to leave last_updated_at as NULL when first connecting. Based on my research today that seems like the right thing, but happy to discuss.

I don't have time to QA today. Let's just QA it on staging...

File: spec/features/admin/accounting/settings_flow_spec.rb (hunk 29 above)
Comment (DevneyHamilton) [reply spec/features/admin/accounting/settings_flow_spec.rb:87]: hmmm i think adding a test about the job failing is redundant of the task specs, which test that the task correctly reports the status of the job.

File: app/controllers/admin/accounting/quickbooks_controller.rb (hunk 2 above)
Comment (DevneyHamilton) [reply app/controllers/admin/accounting/quickbooks_controller.rb]: thx :)

File: app/controllers/admin/accounting/quickbooks_controller.rb (hunk 5 above)
Comment (DevneyHamilton) [reply app/controllers/admin/accounting/quickbooks_controller.rb]: yeah that would be bad. i'm adding a simple error raise & if it needs to be different pls let me know.

File: spec/features/admin/accounting/settings_flow_spec.rb (hunk 29 above)
Comment (smoyte) [reply spec/features/admin/accounting/settings_flow_spec.rb:87]: Good point. The only thing not tested then would be that the i18n string matching the status exists I think. But we can let that go. :)

**Pull request merged**